docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,766 @@
+
+This is useful for Android devices that do not support background services or that often kill the 
+background service.
+
+iOS only allows certain kind of applications to keep a connection open when in the
+background, VoIP for example, which current status client does not qualify for.
+
+Applications on iOS can also request execution time when they are in the [background]
+(https://developer.apple.com/documentation/uikit/app_and_environment/scenes/preparing_your_ui_to_run_in_the_background/updating_your_app_with_background_app_refresh)
+but it has a limited set of use cases, for example it won't schedule any time
+if the application was force quit, and generally is not responsive enough to implement
+a push notification system.
+
+Therefore Status provides a set of Push notification services that can be used
+to achieve this functionality.
+
+Because this can't be safely implemented in a privacy preserving manner, clients
+MUST be given an option not to enable push notifications and not to send push notifications.
+
+## Requirements
+
+The party releasing the app needs to run their own [gorush](https://github.com/appleboy/gorush) 
+publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.
+
+## Components
+
+### Gorush instance
+
+A [gorush](https://github.com/appleboy/gorush) instance MUST be publicly 
+available, this will be used only by push notification servers.
+
+### Push notification server
+
+A push notification server used by clients to register for push notification and
+sending push notifications.
+
+### Registering client
+
+A status client that wants to receive push notifications
+
+### Sending client
+
+A status client that wants to send push notifications
+
+## Registering with the push notification service
+
+A client MAY register with a Push Notification service of their choice.
+
+[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)
+
+A message MUST be sent to the [partitioned topic](#partitioned-topic) for the
+public key of the node, encrypted with this key.
+
+The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):
+
+```protobuf
+
+message PushNotificationOptions {
+  boolean enabled = 1;
+  repeated string allowed_user_list = 2;
+  repeated string block_user_list = 3;
+  repeated string block_chat_list = 4;
+}
+
+message PushNotificationDeviceToken {
+  enum TokenType {
+    UNKNOWN_TOKEN_TYPE = 0;
+    APN_TOKEN = 1;
+    FIREBASE_TOKEN = 2;
+  }
+  TokenType token_type = 1;
+  string token = 2;
+  string installation_id = 3;
+  PushNotificationPreferences preferences = 4;
+}
+
+message PushNotificationPreferences {
+  repeated DeviceToken device_tokens = 1;
+  uint version = 2;
+  boolean unregister = 3;
+}
+
+message PushNotificationRegister {
```

Comment:
How do we prevent this message to be intercepted and replayed to sign up for other servers? An attacker could do so to compromise privacy, spam notifications and if a cost is attached drain user funds

Reply:
Good question, so:

1) the message is sent to the server encrypted with their public key, so only the server is able to  access it. 

2) If replied to the same server the actions taken are idempotent.
3)  If you look at `PushNotificationRegister` definition, we `Sign(payload+publickeyoftheserver)` so that any server receiving this message will discard it if not for them.

Reply:
Okay so with these assumptions replaying is only possible if an eavesdropper waits for a user to register and register.

To me the question remains if a server would still discard a notifications sign up if a payment is involved and potentially claimable with this message.

Reply:
> Okay so with these assumptions replaying is only possible if an eavesdropper waits for a user to register and register.

Not sure I follow,
you can only reply this to the same server, as the signature would fail to verify and it would be discarded. 
Replying to the same server will not yield to any result, as the message would be discarded if already processed (it's basically idempotent), due to a `version` inside the message that is part of the hash to be signed, which effectively acts as a `nonce`.

Given the above, is this still replayable in any meaningful sense?

> To me the question remains if a server would still discard a notifications sign up if a payment is involved and potentially claimable with this message.

What do you mean exactly? You mean that someone could offer money to a malicious server to drop the message/reveal the message?

Reply:
Could you elaborate on how the version could replace a nonce? I was thinking about case: user registers, user unregisters, attacker replays to register again.

For the case of malicious servers I was referring to the notifications as a service case (I didn't find specs for that but iirc it was somehow a longterm plan) , depending on the implementation, if the payment is part of the message payload a replay to a malicious server could be used to steal user funds

Reply:
>  I was thinking about case: user registers, user unregisters, attacker replays to register again.

In this case it won't work, `version` is basically the same as an ethereum nonce (minus the fact that they need to be strictly incrementing by one), basically the server will only process requests that have a version strictly greater than the previous one.
So in your case, user will register with a version of `1`. They will unregister, as per spec this will need to have a version of `2`. The server will keep the last version (`2`) stored, as per spec:
```
The server MUST remove all data about this user if `unregistering` is `true`,
apart from the `hash` of the public key and the `version` of the last options,
in order to make sure that old messages are not processed.
```
If `1` (or `2`) is replayed, the server will ignore the message, as the `version` is not strictly greater than `2`

Reply:
I think this is more of a `version` than a `nonce`, a `nonce` is literally a `number used once`.
In our case 1) we always refer to the same data-structure, which clients are supposed to pull from the server and update 2) The nonce needs to be monotonically increasing 

I guess it depends on what people are most familiar with, but this is an established pattern by distributed databases using optimistic locking and generally version is used ( elasticsearch/dynamodb)